### PR TITLE
Reduce allocations when working with ServiceGroup

### PR DESCRIPTION
**Pull request by reset**
ServiceGroup now contains a single string allocation and three
functions to access parts of the service group. This allows us
to greatly reduce the amount of string allocations when attempting
to comopare the full representation of a service group instead of
it's components.

I also did a bit of work refactoring the public API of the key
generation functions to make them easier to use

gif-keyboard-8803244473903942467

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/hab-butterfly/src/cli.rs [hunk 1]
```diff
@@ -87,10 +87,8 @@ fn file_exists_or_stdin(val: String) -> result::Result<(), String> {
...
 fn valid_service_group(val: String) -> result::Result<(), String> {
-    let regex = Regex::new(r"([A-Za-z_0-9]+)\.([A-Za-z_0-9]+)").unwrap();
-    if regex.is_match(&val) {
-        Ok(())
-    } else {
-        Err(format!("SERVICE_GROUP: '{}' is invalid", &val))
+    match ServiceGroup::validate(&val) {
+        Ok(()) => Ok(()),
+        Err(err) => Err(err.to_string()),
     }
```
Comment (mwrock) [inline components/hab-butterfly/src/cli.rs]: very nice.

File: components/core/src/serde_types.in.rs [hunk 2]
```diff
@@ -61,11 +61,7 @@ pub mod service {
     // `service.rs`
 
     #[derive(Clone, Debug, Eq, Hash, PartialEq, Deserialize, Serialize)]
-    pub struct ServiceGroup {
-        pub service: String,
-        pub group: String,
-        pub organization: Option<String>,
-    }
+    pub struct ServiceGroup(pub String);
 }
```
Comment (mwrock) [inline components/core/src/serde_types.in.rs]: Oh. I see you are using a Jamie style struct!

**Comment by mwrock** [comment]
This looks great! Very nice not to have to `.to_string()` everything too!

**Comment by mwrock** [comment]
@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/196572462).

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/196572462).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

**Pull request merged**